### PR TITLE
Use full package path for definition name in OpenAPI spec

### PR DESCRIPTION
**Pull request by mbohlool**
We were using short package name (last part of package name) plus type name for OpenAPI spec definition name. That can result in duplicate names and make the spec invalid. To be sure we will always have unique names, we are going to use full package name as definition name. Also "x-kubernetes-tag" custom field is added to definitions to list Group/Version/Kind for the definitions that has it. This will help clients to discover definitions easier.
Lastly, we've added a reference from old definition names to the new ones to keep backward compatibilities. The list of old definitions will not be updated.

**Release note**:
- Rename OpenAPI definition names to type's full package names to prevent duplicates
- Create OpenAPI extension "x-kubernetes-group-version-kind" for definitions to store Group/Version/Kind
- Deprecate old definition names and create a reference to the new definitions. Old definitions will be removed in the next release.

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 1]
```diff
@@ -575,6 +575,24 @@ func (s *Scheme) generateConvertMeta(in interface{}) (conversion.FieldMatchingFl
...
+func (s *Scheme) ExportTypesToGVKMaps() map[reflect.Type][]schema.GroupVersionKind {
+	ret := map[reflect.Type][]schema.GroupVersionKind{}
+	for k, v := range s.typeToGVK {
+		ret[k] = v
+	}
+	return ret
+}
+
+// Creates and returns a copy of Scheme.unversionedTypes
+func (s *Scheme) ExportUnversionedTypesMap() map[reflect.Type]schema.GroupVersionKind {
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: I don't think we need to add these methods since you can already get them via traversal?

File: pkg/genericapiserver/server/openapi/openapi.go [hunk 2]
```diff
@@ -79,6 +81,15 @@ func (o *openAPI) init(webServices []*restful.WebService) error {
...
 		}
 	}
+	if o.config.GetDefinitionName == nil {
+		o.config.GetDefinitionName = func(_, name string) (string, []string) {
+			return name[strings.LastIndex(name, "/")+1:], nil
+		}
+	}
+	o.definitions = o.config.GetDefinitions(func(name string) spec.Ref {
+		defName, _ := o.config.GetDefinitionName(o.servePath, name)
+		return spec.MustCreateRef("#/definitions/" + openapi.EscapeJsonPointer(defName))
```
Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: Even with escaping i'm pretty sure this is going to break naive clients.

File: pkg/genericapiserver/server/openapi/openapi.go [hunk 3]
```diff
@@ -33,13 +33,15 @@ import (
 
 const (
 	OpenAPIVersion = "2.0"
+	ExtensionTags  = "x-kubernetes-tags"
```
Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: Create a specific name for each extension, rather than a generic one.  I don't think this name describes what we're exposing well

Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: Sure, I am going to use x-kubernetes-gvk and make the value a little more descriptive.

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 2 above)
Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: can you elaborate?

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go (hunk 1 above)
Comment (mbohlool) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: I am not sure how?

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: I would not use acronyms in the extension name - should be "group-version-kind" or "kind"

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: We should verify that this doesn't horribly break any of the generation code we use today (docs, client code) - I am skeptical that people implementing those libraries handle escaping correctly, because no one handles escaping correctly.

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go (hunk 1 above)
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: AllKnownTypes, New, and IsUnversioned will provide this info.

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 3 above)
Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: I've changed it to x-kubernetes-group-version-kind and also change the format of the content to specify which one is group, version, and kind.

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 2 above)
Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: The escaping is a standard which without it, the spec would be invalid (the validation will fail for it). The client and doc generators are already doing some preprocessing before consuming this spec, and even without this escaping we need to do more preprocessing to use this new spec. Adding a step to deal with this in the preprocessing shouldn't be hard (if the standard json readers does not deal with it).

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go (hunk 1 above)
Comment (mbohlool) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: Good point. Fixed.

**Comment by mbohlool** [comment]
@smarterclayton I've added fixup commits to address your comments. I didn't rebase so you can see the changes. After LGTM, I will rebase to fix the conflicts as well as squash fixup commits.

File: pkg/genericapiserver/server/openapi/openapi.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: I would just like to know if we break the spec just by adding this escaping for any of our existing tools before we merge it.

Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: There are two tools that are currently using this. The documentation (cc @pwittrock) and python-client. For the python client, with no pre-processing, it will break the client generator and we need to update pre-processing step and re-generate it. The other option is not to use "/" as separator.

Comment (smarterclayton) [inline pkg/genericapiserver/server/openapi/openapi.go]: "." Would be the most used separator in the widest range of languages.  If we do that I think we should also do domain reversing (github.com -> com.github).  Go's type naming systems is fairly obscure.

Comment (mbohlool) [inline pkg/genericapiserver/server/openapi/openapi.go]: Done.

**Comment by mbohlool** [comment]
@mikedanese Bazel test changes the package name when running:

```bash
bazel test //pkg/genericapiserver/server/openapi:go_default_test
 ```

The package name for types inside that test file should be ...openapi.TestOutput etc. but it is ...openapi.go_default_test.TestOutput. How can I fix it or disable bazel test for this target?

**Comment by mbohlool** [comment]
@k8s-bot gce etcd3 e2e test this

File: pkg/genericapiserver/endpoints/openapi/openapi.go [hunk 4]
```diff
@@ -85,3 +94,54 @@ func GetOperationIDAndTags(servePath string, r *restful.Route) (string, []string
...
+		typeGroupVersionKinds: map[string][]v1.GroupVersionKind{},
+	}
+	for gvk, rtype := range api.Scheme.AllKnownTypes() {
+		ret.typeGroupVersionKinds[typeName(rtype)] = append(ret.typeGroupVersionKinds[typeName(rtype)], gvkConvert(gvk))
+	}
+	return ret
+}
+
+// GetDefinitionName returns the name and tags for a given definition
+func (d *DefinitionNamer) GetDefinitionName(servePath string, name string) (string, spec.Extensions) {
```
Comment (smarterclayton) [inline pkg/genericapiserver/endpoints/openapi/openapi.go]: Can you add a test case that makes sure this works correctly in the openapi_test.go group?

Comment (mbohlool) [inline pkg/genericapiserver/endpoints/openapi/openapi.go]: I will add a test for this but not sure what you exactly mean by openapi_test.go "group"?

Comment (mbohlool) [inline pkg/genericapiserver/endpoints/openapi/openapi.go]: done adding a test.

File: pkg/genericapiserver/endpoints/openapi/openapi_test.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"testing"
+
+	"github.com/go-openapi/spec"
+	assert_pkg "github.com/stretchr/testify/assert"
```
Comment (smarterclayton) [inline pkg/genericapiserver/endpoints/openapi/openapi_test.go]: style: Generally we don't use assert unless we have to - it's preferred for tests to not import extra dependencies

Not a huge deal, but I prefer keeping our tests fairly close to Go standard unless there's a big win from the import.

**Comment by smarterclayton** [comment]
Looks good, squash the duplicated commits and then you can self apply.

/approve

File: pkg/genericapiserver/endpoints/openapi/openapi_test.go (hunk 5 above)
Comment (mbohlool) [inline pkg/genericapiserver/endpoints/openapi/openapi_test.go]: done.

**Comment by mbohlool** [comment]
PR needs rebase constantly because it touches all openapi specs and any type change will result in a rebase. I set priority to P1 so it gets merged faster. Let me know if you have any objections.

**Comment by mbohlool** [comment]
@k8s-bot bazel test this

**Comment by mbohlool** [comment]
@smarterclayton can you approve this again please?

**Comment by mbohlool** [comment]
@lavalamp this needs approval again.

**Comment by deads2k** [comment]
It looks like this took a generic package and turned it into a kube specific package.

File: pkg/genericapiserver/endpoints/openapi/openapi.go [hunk 6]
```diff
@@ -85,3 +96,382 @@ func GetOperationIDAndTags(servePath string, r *restful.Route) (string, []string
...
+	if groupVersionKinds, ok := d.typeGroupVersionKinds[name]; ok {
+		return friendlyName(name), spec.Extensions{
+			extensionGVK: []v1.GroupVersionKind(groupVersionKinds),
+		}
+	}
+	return friendlyName(name), nil
+}
+
+// PostProcessSpec finalize OpenAPI spec and add removed definition for backward compatibility
+func PostProcessSpec(s *spec.Swagger) (*spec.Swagger, error) {
```
Comment (deads2k) [inline pkg/genericapiserver/endpoints/openapi/openapi.go]: @lavalamp @mbohlool All this is non-generic. It's specific for a kube-apiserver.  Why was this added to a package in genericapiserver?

Comment (mbohlool) [inline pkg/genericapiserver/endpoints/openapi/openapi.go]: https://github.com/kubernetes/kubernetes/pull/40829 sent to fix this.
